llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp

```diff
@@ -448,6 +448,20 @@ Instruction *InstCombinerImpl::visitMul(BinaryOperator &I) {
   if (match(Op1, m_ZExt(m_Value(X))) && X->getType()->isIntOrIntVectorTy(1))
     return SelectInst::Create(X, Op0, ConstantInt::getNullValue(Ty));
 
+  // mul (sext X), Y -> select X, -Y, 0
+  // mul Y, (sext X) -> select X, -Y, 0
+  Value *SExtOp;
+  if (match(Op0, m_OneUse(m_SExt(m_Value(SExtOp)))) &&
```

Comment:
Please add the following tests: https://godbolt.org/z/9856M5M8d
```
define i32 @test_mul_sext_bool(i1 %x, i32 %y) {
  %sext = sext i1 %x to i32
  %mul = mul i32 %sext, %y
  ret i32 %mul
}

define i32 @test_mul_sext_bool_nuw(i1 %x, i32 %y) {
  %sext = sext i1 %x to i32
  %mul = mul nuw i32 %sext, %y
  ret i32 %mul
}

define i32 @test_mul_sext_bool_nsw(i1 %x, i32 %y) {
  %sext = sext i1 %x to i32
  %mul = mul nsw i32 %sext, %y
  ret i32 %mul
}

define i32 @test_mul_sext_bool_nuw_nsw(i1 %x, i32 %y) {
  %sext = sext i1 %x to i32
  %mul = mul nuw nsw i32 %sext, %y
  ret i32 %mul
}

define i32 @test_mul_sext_bool_commuted(i1 %x, i32 %y) {
  %yy = xor i32 %y, 1 ; thwart complexity-based canonicalization
  %sext = sext i1 %x to i32
  %mul = mul i32 %yy, %sext
  ret i32 %mul
}

define i32 @test_mul_sext_nonbool(i2 %x, i32 %y) {
  %sext = sext i2 %x to i32
  %mul = mul i32 %sext, %y
  ret i32 %mul
}

define i32 @test_mul_sext_multiuse(i1 %x, i32 %y) {
  %sext = sext i1 %x to i32
  call void @use(i32 %sext)
  %mul = mul i32 %sext, %y
  ret i32 %mul
}

declare void @use(i32 %x)
```

Reply:
```llvm
define i32 @test_mul_sext_bool_commuted(i1 %x, i32 %y) {
; CHECK-LABEL: @test_mul_sext_bool_commuted(
; CHECK-NEXT:    [[TMP1:%.*]] = xor i32 [[Y:%.*]], -2
; CHECK-NEXT:    [[YY_NEG1:%.*]] = add i32 [[TMP1]], 1
; CHECK-NEXT:    [[MUL:%.*]] = select i1 [[X:%.*]], i32 [[YY_NEG1]], i32 0
; CHECK-NEXT:    ret i32 [[MUL]]
;
  %yy = xor i32 %y, 1
  %sext = sext i1 %x to i32
  %mul = mul i32 %yy, %sext
  ret i32 %mul
}
```
seems incorrect?

Reply:
It is correct. Alive2: https://alive2.llvm.org/ce/z/sTPbP3